core/services/workflows/engine.go

```diff
@@ -615,17 +603,55 @@ func (e *Engine) finishExecution(ctx context.Context, executionID string, status
 		return err
 	}
 
+	e.stepUpdatesChMapMu.Lock()
```

Comment:
This looks correct, but I wonder if we're actually leaving some threads hanging atm.

If you look at handleStepUpdate, this function gets called when we early exit (line 547) but there's a comment indicating that we'll continue to execute the other branches, which is correct.

Could you test this out by writing a test for me?

The test should describe a workflow with two branches, eg.

trigger -> consensus -> targetA
                                     -> targetB

could you mock targetA so it blocks until a signal is received and then completes, while targetB should execute immediately but early exit? Then once targetB has finished you can fire the signal to target A. I would expect this to panic.

To early exit, you need to return the special ErrStopExecution from the capability.

You might also need to add an onStepUpdated handler like we have onExecutionFinished so you can avoid sleeps and coordinate the test correctly.

Lmk if that makes sense